documentation/projects/proposals/document_all_media_properties/20240313-implementation_plan_document_media_properties_api.md

```diff
@@ -0,0 +1,239 @@
+
+Using introspection we can obtain a list of all fields on the model, and then
+extract the following data points for each field.
+
+| Data-point      | Type | Description                                   |
+| --------------- | ---- | --------------------------------------------- |
+| `name`          | str  | the name of the field                         |
+| `internal_type` | str  | the class name of the field                   |
+| `db_type`       | str  | the database type of the field                |
+| `is_relation`   | bool | whether the field is relation to other models |
+
+The following data points are only present (and relevant) if the field is a
+relationship.
+
+| Data-point | Type | Description                                         |
+| ---------- | ---- | --------------------------------------------------- |
+| `nature`   | str  | the one/many to one/many nature of the relationship |
+| `to`       | str  | the name of the related model                       |
+
+The following datapoints are only present (and relevant) if the field is not a
+relationship.
+
+| Data-point     | Type | Description                                    |
+| -------------- | ---- | ---------------------------------------------- |
+| `allows_blank` | bool | whether the field accepts blank strings        |
+| `allows_null`  | bool | whether the field accepts null values          |
+| `is_pk`        | bool | whether the field has a primary key constraint |
+| `needs_unique` | bool | whether the field has a unique index           |
+| `default`      | bool | the value to use if one is not provided        |
+| `help_text`    | str  | a human-readable description of the field      |
+
+[np]:
+  https://docs.djangoproject.com/en/4.0/ref/models/fields/#django.db.models.Field.default
+
+```{note}
+- The `is_relation` data-point is used to render two separate tables for
+  relation and value fields.
+- The `allows_null` data-point is not relevant for relations as they are always
+  nullable.
+- The `max_length` data-point is not captured because it is automatically
+  covered under the `db_type` data-point.
+- `is_pk`, `needs_unique`, `allows_null` and `allows_blank` are presented
+  collectively as "constraints" although not technically DB-level constraints.
+```
+
+The information extracted from this introspection will be stored in memory for
+the next stage.
+
+### Merge with additional notes
+
+Additional notes or hardcoded information may be extra information about a field
+such as known inconsistencies or special cases. This information, if documented
+using a consistent format, can be used to augment the information extracted in
+the previous step.
+
+We will document this information in the docstring of the model classes from
+where it can be accessed using `inspect.getdoc`. To ensure that these notes are
+co-located with the fields' code, we will parse docstrings of all ancestors and
+mixins of a model class.
```

Comment:
Will this work with `help_text` that is located in a separate file and is added inside `__init__`, as the `source` ones?
https://github.com/WordPress/openverse/blob/12d7162fb9a9c44a019ea329d87f573447792bc8/api/api/serializers/media_serializers.py#L310-L311

Reply:
These help texts are taken from the model and not from the serializer. That's because for a specific model a number of serializers may be defined that can have different `help_text`s as per the requirements of the endpoint.

So for source, we show this

https://github.com/WordPress/openverse/blob/12d7162fb9a9c44a019ea329d87f573447792bc8/api/api/models/media.py#L56-L58

Reply:
You can visit https://docs.openverse.org/_preview/3966/meta/media_properties/api.html to see the the generated docs output.